### PR TITLE
Require minimum C++11 and switch --with-cpp11 to --with-cpp14 (continuation)

### PR DESCRIPTION
**Pull request by rouault**
This is a continuation of https://github.com/OSGeo/gdal/pull/243, rebased on top of current trunk (which has a few source required code changes as well)

Issues with C++11 and Python have been fixed by detecting if a -std=c++11 flag is needed to compile C++11 code, and to add it in that case

For Perl, I've hastily done a hack, to add -std=c++11, which @ajolma should review. Another unsatisfactory aspect for Perl bindings is that to compile the generated .cpp files, it seems that the compiler name seems to be hardcoded ('cc' or 'x86_64-linux-gnu-gcc') and CC or CXX environment variables are not honoured, which prevent from building the Perl bindings on a platform where the default compiler would not be C++11 capable, whereas an alternate compiler would be. I guess that can be tuned ?

Other various updates in configurations to switch to distributions/compilers supporting C++11

### CODE REVIEW

File: gdal/swig/perl/Makefile.PL [hunk 1]
```diff
@@ -53,7 +53,7 @@ for my $module (sort keys %$objects) {
     my $LD = $Config{ld};
     $LD .= ' '.$ENV{CFLAGS} if $ENV{CFLAGS};
     $LD .= ' '.$ENV{LDFLAGS} if $ENV{LDFLAGS};
-    my $OPTIMIZE = '';
+    my $OPTIMIZE = '-std=c++11';
```
Comment (rouault) [inline gdal/swig/perl/Makefile.PL]: @ajolma Would there be a way of dynamically checking if we need to add this flag ? Something in spirit similar to the Python equivalent in https://github.com/OSGeo/gdal/pull/270/commits/e2d6633eed23514b1b61d8fa52b64b1e58f4576d

File: gdal/ci/travis/gcc48_stdcpp11/install.sh [hunk 2]
```diff
@@ -18,10 +18,10 @@ cat java.opt | sed "s/JAVA_HOME =.*/JAVA_HOME = \/usr\/lib\/jvm\/java-7-openjdk-
 mv java.opt.tmp java.opt
 make
 cd ../..
-cd swig/perl
-make generate
-make
-cd ../..
+#cd swig/perl
+#CC=g++-4.8 make generate
```
Comment (rouault) [inline gdal/ci/travis/gcc48_stdcpp11/install.sh]: @ajolma would there be a way of making this work (ie forcing an alternate compiler than the default one) ?

File: gdal/swig/perl/Makefile.PL (hunk 1 above)
Comment (rouault) [reply gdal/swig/perl/Makefile.PL]: Or perhaps use the CXX variable from GDALmake.opt (which has -std=c++11 if needed)

File: gdal/port/cpl_port.h [hunk 3]
```diff
@@ -179,11 +179,17 @@
...
-#  if __cplusplus >= 201103L
-#    define HAVE_CXX11 1
+#  if !(__cplusplus >= 201103L || _MSC_VER >= 1800)
+#    error Must have C++11 or newer.
+#  endif
+#  if __cplusplus >= 201402L
+#    define HAVE_CXX14 1
 #  endif
-/* TODO(schwehr): What are the correct tests for C++ 14 and 17? */
+/* TODO(schwehr): What are the correct test for C++ 17? */
```
Comment (schwehr) [inline gdal/port/cpl_port.h]: are -> is

File: gdal/ci/travis/gcc48_stdcpp11/install.sh (hunk 2 above)
Comment (ajolma) [reply gdal/ci/travis/gcc48_stdcpp11/install.sh]: Probably Makefile.PL should be edited to pick the selected compiler from GDALmake.opt. Is that CXX (without the $(LIBTOOL_COMPILE_CXX)? And put it into the call to WriteMakefile as "CC => $compiler,". I'm already going through GDALmake.opt in Makefile.PL to get the man dirs.

File: gdal/swig/perl/Makefile.PL (hunk 1 above)
Comment (ajolma) [reply gdal/swig/perl/Makefile.PL]: See above, I guess. We can put the switch into CC => 'g++ -std=c++11' (actually in a variable that is picked from the opt.

**Comment by ajolma** [comment]
I would do this:
```
Index: Makefile.PL
===================================================================
--- Makefile.PL	(revision 40921)
+++ Makefile.PL	(working copy)
@@ -20,30 +20,36 @@
         "You have to first say \"cd $source_tree; make\".";
 }
 
-# logic to set INSTALLSITEMAN3DIR if INSTALL_BASE and INST_MAN are set
+# Obtain some information from GDALmake.opt.
+#
+# 1) Set INSTALLSITEMAN3DIR if INSTALL_BASE and INST_MAN are set.
 # INSTALL_BASE is used in the standard case when this script is called
 # as a part of overall build with --with-perl
 # INST_MAN may be set by --mandir=DIR option to configure
-if ($ARGV[0] =~ /INSTALL_BASE=(.*)/) {
-    my $INSTALL_BASE = $1;
-    my $opt = "$source_tree/GDALmake.opt";
-    open(my $fh, "<", $opt) or die "Can't open < $opt: $!";
-    while (<$fh>) {
-        if (/(\w+)\s+=\s+(.*)/) {
-            my $key = $1;
-            my $val = $2;
-            $val =~ s/\s+$//;
-            if ($key eq 'prefix') {
-                die 'prefix in GDALmake.opt is set different to \$INSTALL_BASE' unless $INSTALL_BASE eq $val;
-            } elsif ($key eq 'INST_MAN') {
-                $val =~ s/\$\{prefix\}/$INSTALL_BASE/;
-                push @ARGV, "INSTALLSITEMAN1DIR=$val/man1";
-                push @ARGV, "INSTALLSITEMAN3DIR=$val/man3";
-            }
+#
+# 2) The C++ compiler.
+#
+my $cxx = '';
+my $opt = "$source_tree/GDALmake.opt";
+open(my $fh, "<", $opt) or die "Can't open < $opt: $!";
+my ($INSTALL_BASE) = $ARGV[0] =~ /INSTALL_BASE=(.*)/;
+while (<$fh>) {
+    if (/(\w+)\s+=\s+(.*)/) {
+        my $key = $1;
+        my $val = $2;
+        $val =~ s/\s+$//;
+        if ($key eq 'prefix') {
+            die 'prefix in GDALmake.opt is different from \$INSTALL_BASE' unless $INSTALL_BASE eq $val;
+        } elsif ($key eq 'INST_MAN') {
+            $val =~ s/\$\{prefix\}/$INSTALL_BASE/;
+            push @ARGV, "INSTALLSITEMAN1DIR=$val/man1";
+            push @ARGV, "INSTALLSITEMAN3DIR=$val/man3";
+        } elsif ($key eq 'CXX') {
+            ($cxx) = $val =~ /\)\s*(.*)$/;
         }
     }
-    close $fh;
 }
+close $fh;
 
 my ($INC, $LIB, $objects) = get_config($config);
 
@@ -69,6 +75,7 @@
                    ABSTRACT => 'Perl extension for the GDAL library for geospatial data',
                    AUTHOR => 'Ari Jolma <ari.jolma at gmail.com>',
                    MAKEFILE => 'Makefile_'.$add,
+                   CC => $cxx,
                    LIBS => $LIB,
                    INC => $INC,
                    OPTIMIZE => $OPTIMIZE,
```

**Comment by rouault** [comment]
Ari, I believe I'll let you do any refinement in Perl bindings once this is committed. By the way, is it possible to build Perl bindings on Windows (is MSVC supported ? or just mingw/cygwin ?). Asking that since my currently hardcoded -std=c++11 wouldn't please MSVC

**Comment by ajolma** [comment]
I haven't build Perl on Windows for ~10 years but looking at Internet both may/seem to be supported. I'll bring this up next week with someone I believe is more experienced.

**Comment by rouault** [comment]
Committed in r40984

**Pull request closed without merging**